---
title: "Installation of RSAT on an Ubuntu device"
output:
  pdf_document:
    toc: yes
  html_document:
    css: /packages/rsat/public_html/course.css
    fig_caption: yes
    toc: yes
---
**/!\\ NB** If you are trying to install RSAT on an IFB appliance, it is assumed that it has a root account and a standard user account with granted sudoer privileges (see IFB_cloud_VM_creation.Rmd for details).

Unless stated otherwise, the following commands will be executed as root. 

**/!\\ TODO** Check this protocole with VirtualBox appliance.

## 1. First steps

### Set the locales manually if needed.
```
export LANGUAGE=en_US.UTF-8
export LANG=en_US.UTF-8
export LC_ALL=en_US.UTF-8
locale-gen en_US.UTF-8
sudo dpkg-reconfigure locales
```

### Check date & time, and adjust your timezone if needed. 
```
date
```
```
dpkg-reconfigure tzdata
```

### Set  installation parameters.
<!---
```
export INSTALLER=apt-get
export INSTALLER_OPT="--quiet --assume-yes"
```
-->

<!---
We need to update apt-get, to avoid trouble with python
See http://askubuntu.com/questions/350312/i-am-not-able-to-install-easy-install-in-my-ubuntu
-->

### Create RSAT directory. It must be readable by all users (in particular by the apache user).
```
mkdir -p /packages
cd /packages
mkdir -p /packages/install_logs
chmod 777 /packages/install_logs
df -m > /packages/install_logs/df_$(date +%Y-%m-%d_%H-%M-%S)_start.txt
```

### Check the installation device. This should give sda1 or vda1?
```
DEVICE=`df -h / | grep '/dev'| awk '{print $1}' | perl -pe 's/\/dev\///'`
echo ${DEVICE}
```
        
## 2. Packages installation

### You can check the increase of disk usage during the different steps of the installation.
```
grep ${DEVICE} /packages/install_logs/df_*.txt
```

### Update apt-get.
```
apt-get update ## Problem with GRUB?
df -m > /packages/install_logs/df_$(date +%Y-%m-%d_%H-%M-%S)_apt-get_updated.txt
apt-get --quiet --assume-yes upgrade
df -m > /packages/install_logs/df_$(date +%Y-%m-%d_%H-%M-%S)_apt-get_upgraded.txt
grep ${DEVICE} /packages/install_logs/df_*.txt
```

### Packages to be installed. **/!\\** Are those all indispensable? Not already installed?
<!---
To be removed?
```
PACKAGES_OPT="
ess
yum
php-elisp
libgdbm-dev
libgd2-xpm-dev
libxml2-dev
links
gfortran
libmysqlclient-dev
texlive-latex-base
python-virtualenv
ipython
ipython-notebook
libssl-dev
libreadline-gplv2-dev:i386
lib64readline-gplv2-dev:i386
libreadline-gplv2-dev
libx11-dev
libxt-dev
libxml2-dev
tcl8.5-dev
tk8.5-dev
libxss-dev
libpng12-dev
libjpeg62-dev
libcairo2-dev
lib32z1
lib32ncurses5
lib32bz2-1.0
libc6-dev
build-essential
python-dev
python3-dev
libnet-ssleay-perl
libcrypt-ssleay-perl
exfat-fuse
exfat-utils 
at
"
```
-->
```
PACKAGES="
ssh
git
cvs
wget
zip
unzip
finger
screen
make
g++
apache2
php5
libapache2-mod-php5
libgd-tools
libgd-gd2-perl
ghostscript
gnuplot
graphviz
mysql-client
default-jre
python
python-pip
python-setuptools 
python-numpy
python-scipy
python-matplotlib
python-suds
python3
python3-pip
python3-setuptools 
python3-numpy
python3-scipy
python3-matplotlib
r-base
emacs
x11-apps
firefox
eog
ntp
curl
libcurl4-openssl-dev
"
```
### Perl modules installation (could be done with cpan, but ensure consistency with Ubuntu OS?)
```
PACKAGES_PERL="perl-doc
pmtools
libyaml-perl
libemail-simple-perl
libemail-sender-perl
libemail-simple-creator-perl
libpostscript-simple-perl
libstatistics-distributions-perl
libio-all-perl
libobject-insideout-perl
libobject-insideout-perl
libsoap-lite-perl
libsoap-wsdl-perl
libxml-perl
libxml-simple-perl
libxml-compile-cache-perl
libdbi-perl
liblockfile-simple-perl
libobject-insideout-perl
libgd-perl
libdbd-mysql-perl
libjson-perl
libbio-perl-perl
libdigest-md5-file-perl
libnet-address-ip-local-perl
"
```

<!---
Not used later?
### Additional Perl modules not found with apt-get. These will have to be installed with cpan.
```
PACKAGES_PERL_MISSING="
libalgorithm-cluster-perl
digest-md5-file-perl
liblockfile-simple
libutil-properties-perl
librest-client-perl
libxml-compile-soap11-perl
libxml-compile-wsdl11-perl
libxml-compile-transport-soaphttp-perl
libbio-das-perl        
"
```
-->

### Install the apt-get libraries.
```
echo "Packages to be installed with apt-get --quiet --assume-yes"
echo "${PACKAGES}"
echo "Perl module packages to be installed with apt-get --quiet --assume-yes"
echo "${PACKAGES_PERL}"
for LIB in ${PACKAGES} ${PACKAGES_PERL}; \
do \
   echo "`date '+%Y/%m/%d %H:%M:%S'`  installing apt-get library ${LIB}" ; \
   apt-get install --quiet --assume-yes ${LIB} > /packages/install_logs/apt-get_install_${LIB}.txt ; \
   df -m > /packages/install_logs/df_$(date +%Y-%m-%d_%H-%M-%S)_${LIB}_installed.txt ; \
done
echo "Log files are in folder /packages/install_logs"
grep ${DEVICE} /packages/install_logs/df_*.txt
```

### Warning messages: 
```
Warning: there was a problem reading the certificate file /etc/ssl/certs/EBG_Elektronik_Sertifika_Hizmet_Sa?lay?c?s?.pem. Message:
  /etc/ssl/certs/EBG_Elektronik_Sertifika_Hizmet_Sa?lay?c?s?.pem (No such file or directory)
Warning: there was a problem reading the certificate file /etc/ssl/certs/AC_Ra?z_Certic?mara_S.A..pem. Message:
  /etc/ssl/certs/AC_Ra?z_Certic?mara_S.A..pem (No such file or directory)
Warning: there was a problem reading the certificate file /etc/ssl/certs/NetLock_Arany_=Class_Gold=_F?tan?s?tv?ny.pem. Message:
  /etc/ssl/certs/NetLock_Arany_=Class_Gold=_F?tan?s?tv?ny.pem (No such file or directory)
Warning: there was a problem reading the certificate file /etc/ssl/certs/Certinomis_-_Autorit?_Racine.pem. Message:
  /etc/ssl/certs/Certinomis_-_Autorit?_Racine.pem (No such file or directory)
Warning: there was a problem reading the certificate file /etc/ssl/certs/T?B?TAK_UEKAE_K?k_Sertifika_Hizmet_Sa?lay?c?s?_-_S?r?m_3.pem. Message:
  /etc/ssl/certs/T?B?TAK_UEKAE_K?k_Sertifika_Hizmet_Sa?lay?c?s?_-_S?r?m_3.pem (No such file or directory)
```

### Package to be installed in an interactive mode.
```
apt-get install --quiet --assume-yes console-data
```
* Options:
    * Select keymap from arch list
    * \<Don't touch keymap\> (default)
    * Keep kernel keymap
    * Select keymap from full list

### Specific treatment for some Python libraries.
```
sudo apt-get --quiet --assume-yes build-dep python-numpy python-scipy
df -m > /packages/install_logs/df_$(date +%Y-%m-%d_%H-%M-%S)_numpy-scipy_dependencies_installed.txt
grep ${DEVICE} /packages/install_logs/df_*.txt
```
**/!\\** Remove these from ${PACKAGES} then?

### To free space, remove apt-get packages that are no longer required. /?\\
```
grep ${DEVICE} /packages/install_logs/df_*.txt
apt-get --quiet --assume-yes  autoremove
df -m > /packages/install_logs/df_$(date +%Y-%m-%d_%H-%M-%S)_autoremoved.txt
apt-get --quiet --assume-yes  clean
df -m > /packages/install_logs/df_$(date +%Y-%m-%d_%H-%M-%S)_cleaned.txt
grep ${DEVICE} /packages/install_logs/df_*.txt
```

## 3. Python libraries installation

**/!\\** Check whether compatible with libraries used in snakemake (pip3). 

**Note:** I removed commented packages.

**/!\\ TODO:** Check modules that are necessary for RSAT to work, and those which can be installed separately (eg. snakemake?). 

Note: at this stage, numpy, scipy and matplotlib have already been installed with apt-get under Ubuntu. For other OS, they should be added to the pip installation. 

```
pip install soappy
pip install fisher
pip install httplib2
```

Optional: a utility to measure internet bandwidth.
```
sudo pip install speedtest-cli
```

<!---
Some other packages:
```
sudo pip3 install fisher
sudo pip3 install snakemake
```
-->

### Check disk usage
df -m > /packages/install_logs/df_$(date +%Y-%m-%d_%H-%M-%S)_pip_libraries_installed.txt
grep ${DEVICE} /packages/install_logs/df_*.txt

## 4. Apache Web server configuration

**/!\\** Manual interventions needed here. 

### Activate CGI module.
```
vi /etc/apache2/sites-available/000-default.conf
```
Uncomment the following line: `Include conf-available/serve-cgi-bin.conf`.

### To avoid puzzling warning at apache start, set ServerName globally.
```
vi /etc/apache2/apache2.conf
```
Add the following line at the end of the file: `ServerName localhost`.
                                           
### Add CGI script.
```
vi /etc/apache2/mods-available/mime.conf
```
Uncomment the line `AddHandler cgi-script .cgi`.

Optional: associate a plain/text mime type to extensions for some classical bioinformatics files.
`AddType text/plain .fasta`
`AddType text/plain .bed`.

### Adapt the PHP parameters.
```
nano /etc/php5/apache2/php.ini
```
Modify the following parameters: `post_max_size = 100M` and `upload_max_filesize=100M`.

### Activate cgi scripts. Found [here](http://www.techrepublic.com/blog/diy-it-guy/diy-enable-cgi-on-your-apache-server/).
```
chmod 755 /usr/lib/cgi-bin
chown root.root /usr/lib/cgi-bin
a2enmod cgi
service apache2 restart
```
You can check whether apache server was successfully configured and started by opening a web connection to `http://192.54.201.178`.

#### OK til here for IFB VM.

<!---
Removed all SOAP tricky stuff
-->


## 5. RSAT distribution

**/!\\ Note:** The git distribution requires an account at the ENS git server, which is currently only possible for RSAT developing team. In the near future, we may use git also for the end-user distribution. For users who don't have an account on the RSAT git server, the code can be downloaded as a tar archive from the Web site.

### Create RSAT directory.

```
mkdir -p /packages/rsat
cd /packages
export RSAT=/packages/rsat
```

### Git repository cloning.

**/!\\ NB:** Used to be done in $HOME, then moved to $RSAT. No longer needed since user rsat was removed from the protocole.

```
git clone git@depot.biologie.ens.fr:rsat
git config --global user.mail claire.rioualen@inserm.fr
git config --global user.name "reg-genomics VM user"
``` 

### Archive download. /!\\ To be checked.

```
export RSAT_DISTRIB=rsat_2015-06-06.tar.gz
export RSAT_DISTRIB_URL=http://pedagogix-tagc.univ-mrs.fr/download_rsat/${RSAT_DISTRIB}
```
```
sudo wget ${RSAT_DISTRIB_URL}
sudo tar -xpzf ${RSAT_DISTRIB}
sudo rm -f ${RSAT_DISTRIB}
sudo df -m > /packages/install_logs/df_$(date +%Y-%m-%d_%H-%M-%S)_rsat_downloaded.txt
cd ~; ln -fs /packages/rsat rsat
```                      

## 6. RSAT configuration

### Run the configuration script, to specify the environment variables.
```
cd $RSAT
sudo perl perl-scripts/configure_rsat.pl
```
Which options to specify?

<!---
Absolute path to the RSAT package ? [/packages/rsat]
Ready to update config file	/packages/rsat/RSAT_config.props [y/n] (y):
rsat_site [your_server_name] : 
rsat_server_admin [your.mail@your.mail.server] :
RSAT [/packages/rsat] :
group_specificity [None] : 
phylo_tools [0] : 
variations_tools [0] : 
ucsc_tools [0] :
ensembl_tools [0] :
RSAT_BIN [/packages/rsat/bin] :
rsat_tmp [/packages/rsat/public_html/tmp] :
mail_supported [no] : 
smtp [] : 
smtp_sender [] : 
rsat_www [auto] : 
rsat_echo [0] : 
start_time [0] : 
exec_time [0] : 
rsat_ws [http://localhost/rsat/] : 
rsat_img_format [png] : 
QUEUE_MANAGER [batch] : 
CLUSTER_QUEUE [rsat] : 
BATCH_MAIL [a] : 
CLUSTER_SHELL [/bin/bash] : 
QSUB_OPTIONS [] : 
REFSEQ_DIR [/packages/rsat/downloads/ftp.ncbi.nih.gov/genomes/refseq] : 
ensembl_host [ensembldb.ensembl.org] : 
ensembl_rsync [rsync://ftp.ensembl.org/ensembl/pub] : 
ensembl_version [79] : 
ensemblgenomes_version [26] : 
ensembl_version_safe [70] : 
ensembl [/packages/rsat/lib/ensemblgenomes-26-79/ensembl/modules] : 
compara [/packages/rsat/lib/ensemblgenomes-26-79/ensembl-compara/modules] : 
variation [/packages/rsat/lib/ensemblgenomes-26-79/ensembl-variation/modules] : 
neat_supported [1] : 
neat_www_root [http://wwwsup.scmbb.ulb.ac.be/rsat/] : 
neat_ws [http://wwwsup.scmbb.ulb.ac.be/rsat/web_services/RSATWS.wsdl] : 
neat_ws_tmp [http://wwwsup.scmbb.ulb.ac.be/rsat/tmp/] : 
neat_java_ws [http://wwwsup.scmbb.ulb.ac.be/be.ac.ulb.bigre.graphtools.server/wsdl/GraphAlgorithms.wsdl] : 
neat_java_host [http://wwwsup.scmbb.ulb.ac.be/rsat/] : 
tomcat_port [] : 
REA_ROOT [/packages/rsat/contrib/REA] : 
KWALKS_ROOT [/packages/rsat/contrib/kwalks/bin] : 
LOGO_PROGRAM [seqlogo] : 
Ready to update config file	/packages/rsat/RSAT_config.mk [y/n] (y): 
RSAT_SITE [your_server_name] : 
RSAT_SERVER_ADMIN [your.mail@your.mail.server] : 
OS [linux] : 
ARCHITECTURE [x64] : 
PACKAGE_MANAGER [apt-get] : 
UCSC_OS [linux.x86_64] : 
SRC_DIR [${RSAT}/app_sources] : 
SUDO [] : 
RSAT_BIN [/packages/rsat/bin] : 
RSAT_WS [http://localhost/rsat/] : 
QUEUE_MANAGER [batch] : 
CLUSTER_QUEUE [rsat] : 
ENSEMBL_RELEASE [79] : 
ENSEMBLGENOMES_BRANCH [26] : 
Ready to update config file	/packages/rsat/RSAT_config.bashrc [y/n] (y): 
Ready to update config file	/packages/rsat/RSAT_config.conf [y/n] (y): 
-->

### Load the (updated) RSAT environment variables.
```
source RSAT_config.bashrc
```

### Check that the RSAT environment variable has been properly configured.
```
echo ${RSAT}
```

### Initialise RSAT folders
```
make -f makefiles/init_rsat.mk init
```

### /?\\ Not sure why those commands.
```
rsync -ruptvl RSAT_config.bashrc /etc/bash_completion.d/
## ln -fs /packages/rsat/RSAT_config.bashrc /etc/bash_completion.d/
emacs -nw /etc/bash.bashrc
```

## 7. Perl modules for RSAT

**/!\\ TODO** Update the list of modules to be installed?

**/!\\** The first time we use cpan, we apparently need to open it manually in order to configure and update it.
```
cpan
```
```
cpan> install YAML
cpan> install CPAN 
cpan> reload cpan
cpan> quit
```

### Get the list of Perl modules to be installed.
```
make -f makefiles/install_rsat.mk  perl_modules_list
make -f makefiles/install_rsat.mk perl_modules_check
more check_perl_modules_eval.txt
grep Fail  check_perl_modules_eval.txt
grep -v '^OK'  check_perl_modules_eval.txt | grep -v '^;'
MISSING_PERL_MODULES=`grep -v '^OK'  check_perl_modules_eval.txt | grep -v '^;' | cut -f 2 | xargs`
echo "Missing Perl modules:     ${MISSING_PERL_MODULES}"
```

### Install the missing Perl modules.
```
make -f makefiles/install_rsat.mk perl_modules_install PERL_MODULES="${MISSING_PERL_MODULES}"
```

### Check once more if all required Perl modules have been correctly installed.
```
make -f makefiles/install_rsat.mk perl_modules_check
more check_perl_modules_eval.txt
```
Note: Object::InsideOut always displays "Fail", whereas it is OK during installation.

> Is this necessary ? Makes no difference apparently.
>
> Note: I had to force installation for the some modules, because
> there seem to be some circular dependencies.
>
> grep -v '^OK'  check_perl_modules_eval.txt | grep -v '^;' | grep -v "Object::InsideOut"
> make -f makefiles/install_rsat.mk perl_modules_install_by_force PERL_MODULES_TO_FORCE="`grep -v '^OK'  check_perl_modules_eval.txt | > grep -v '^;' | grep -v Object::InsideOut| cut -f 2 | xargs`"


### Measure remaining disk space.
df -m > /packages/install_logs/df_$(date +%Y-%m-%d_%H-%M-%S)_perl_modules_installed.txt
grep ${DEVICE} /packages/install_logs/df_*.txt


## 8. Configure RSAT web server

> Uh?
>
> Edit the file to replace [RSAT_PARENT_FOLDER] byt the parent directory of the rsat directory.
>
> cd ${RSAT}
> sudo rsync -ruptvl RSAT_config.conf /etc/apache2/sites-enabled/rsat.conf
> apache2ctl restart

### RSAT Web server URL
```
echo $RSAT_WWW
```

### If the value is "auto", get the URL as follows:
```
export IP=`ifconfig eth0 | awk '/inet /{print $2}' | cut -f2 -d':'`
echo ${IP}
export RSAT_WWW=http://${IP}/rsat/
echo $RSAT_WWW
```

## 9. Other?

### compile RSAT programs written in C
```
make -f makefiles/init_rsat.mk compile_all
export INSTALL_ROOT_DIR=/packages/
df -m > /packages/install_logs/df_$(date +%Y-%m-%d_%H-%M-%S)_rsat_app_compiled.txt
```
**/!\\** Warnings.

### Install some third-party programs required by some RSAT scripts.
```
make -f makefiles/install_software.mk install_ext_apps
df -m > /packages/install_logs/df_$(date +%Y-%m-%d_%H-%M-%S)_rsat_extapp_installed.txt
```

# TODO from here

## 10. Data management

### IFB cloud: link to separate disk?

How exactly are IFB appliances & vdisks linked?

**/?\\** Tree.
$RSAT/data -> $RSAT/public_html
$RSAT/public_html/data

```
export RSAT_DATA_DIR=/root/mydisk/rsat_data
cd ${RSAT}/public_html
mv data/* ${RSAT_DATA_DIR}/
mv data/.htaccess ${RSAT_DATA_DIR}/
rmdir data
ln -s ${RSAT_DATA_DIR} data
cd $RSAT
```

### Install two model organisms, required for some of the Web tools.
```
download-organism -v 1 -org Saccharomyces_cerevisiae -org Escherichia_coli_K_12_substr__MG1655_uid57779
download-organism -v 1 -org Drosophila_melanogaster
download-organism -v 1 -org Caenorhabditis_elegans
download-organism -v 1 -org Arabidopsis_thaliana
```

### Get the list of organisms supported on your computer.
```
supported-organisms
```

### Check again the disk space.
```
sudo df -m > /packages/install_logs/df_$(date +%Y-%m-%d_%H-%M-%S)_rsat_organism_installed.txt
```

## 11. Install selected R librairies

### Packages required for some RSAT scripts.

```
cd $RSAT
make -f makefiles/install_rsat.mk install_r_packages
```
**/?\\** or 
```
cd $RSAT; make -f makefiles/install_rsat.mk update ## install R packages + compile the C programs
```

Done first...

<!---
Not sure what to do with this?
# ## The command R CMD INSTALL apparently does not work at this stage.
# ##	root@rsat-tagc:/workspace/rsat# R CMD INSTALL reshape
# ##	Warning: invalid package 'reshape'
# ##	Error: ERROR: no packages specified
# cd $RSAT; make -f makefiles/install_rsat.mk  r_modules_list 
# ### I install them from the R interface. This should be revised to
# ### make it from the bash, but I need to see how to specify the CRAN
# ### server from the command line (for the time being, I run R and the
# ### programm asks me to specify my preferred CRAN repository the first
# ### time I install packages).
# R
# ## At the R prompt, type the following R commands.
# ##
# ## Beware, the first installation of bioconductor may take a while,
# ## because there are many packages to install
# ##
# ## Note: since this is the first time you install R packages on this
# ## VM, you need to choose a RCRAN server nearby to your site.
# install.packages(c("reshape", "RJSONIO", "plyr", "dendroextras", "dendextend"))
# source('http://packagesconductor.org/packagescLite.R'); biocLite("ctc")
# quit()
# ## At prompt "Save workspace image? [y/n/c]:", answer "n"
-->


### Check remaining disk space
df -m > /packages/install_logs/df_$(date +%Y-%m-%d_%H-%M-%S)_R_packages_installed.txt
grep ${DEVICE} /packages/install_logs/df_*.txt


## 12. Testing RSAT & external programs

### Test a simple Perl script that does not require for organisms to be installed.
```
which random-seq
random-seq -l 100
```

### Test a simple python script that does not require organisms to be installed.
```
random-motif -l 10 -c 0.90
```

### vmatch (used in purge-sequence) **/!\\ TODO** Install mkvtree -> licence ?
```
random-seq -l 100 | purge-sequence
```

### seqlogo **/!\\ TODO**
```
which seqlogo
seqlogo
```

### weblogo 3 **/!\\ TODO**
```
which weblogo
weblogo --help
```

### ghostscript
```
which gs
gs --version
```

### Check that the model genomes have been correctly installed **/!\\ TODO** once data/ properly handled. 
```
# Retrieve all the start codons and count oligonucleotide frequencies (most should be ATG).
retrieve-seq -org Saccharomyces_cerevisiae -all -from 0 -to +2 | oligo-analysis -l 3 -1str -return occ,freq -sort
```

## 13. Configure the SOAP/WSDL Web services

Check the URL of the web services (RSAT_WS). By default, the server addresses the WS requests to itself (http://localhost/rsat) because web services are used for multi-tierd architecture of some Web tools (retrieve-ensembl-seq, NeAT).
```
cd $RSAT
#echo $RSAT_WS
```

### Get the current IP address
```
export IP=`/sbin/ifconfig eth0 | awk '/inet /{print $2}' | cut -f2 -d':'`
echo ${IP}
export  RSAT_WS=http://${IP}/rsat/
```

### Initialize the Web services stub. 
```
make -f makefiles/init_rsat.mk ws_init RSAT_WS=${RSAT_WS}
```

### After this, re-generate the web services stubb, with the following command.
```
make -f makefiles/init_rsat.mk ws_stub RSAT_WS=${RSAT_WS}
```

### Test the local web services
```
make -f makefiles/init_rsat.mk ws_stub_test
```

### Test RSAT Web services (local and remote) without using the SOAP/WSDL stubb (direct parsing of the remote WSDL file)
```
make -f makefiles/init_rsat.mk ws_nostub_test
```

### Test the program supported-organisms-server, which relies on Web services without stub
```
supported-organisms-server -url ${RSAT_WS} | wc
supported-organisms-server -url http://localhost/rsat/ | wc
supported-organisms-server -url http://rsat-tagc.univ-mrs.fr/ | wc
```

### Tests on the Web site

Run the demo of the following tools

* retrieve-seq to check the access to local genomes (at least Saccharomyces cerevisiae)
* feature-map to check the GD library
* retrieve-ensembl-seq to check the interface to Ensembl
* fetch-sequences to ceck the interface to UCSC
* some NeAT tools (they rely on web services)
* peak-motifs because it mobilises half of the RSAT tools -> a good control for the overall installation.
* footprint-discovery to check the tools depending on homology tables (blast tables).

## 14. Install the cluster management system (torque, qsub, ...)

### Check the number of core (processors)
```
grep ^processor /proc/cpuinfo
```

### Check RAM
```
grep MemTotal /proc/meminfo
```


### Install Sun Grid Engine (SGE) job scheduler

### Beware, before installing the grid engine we need to modify manually the file `/etc/hosts`
```
nano /etc/hosts
```

Initial config (problematic)

    127.0.0.1       localhost       rsat-vm-2015-02
    127.0.1.1      rsat-vm-2015-02
    
Config to obtain: 

    127.0.0.1       localhost       rsat-vm-2015-02
    127.0.1.1      rsat-vm-2015-02
    
**/?\\**

```
apt-get install --quiet --assume-yes gridengine-client
apt-get install --quiet --assume-yes gridengine-exec
apt-get install --quiet --assume-yes gridengine-master
apt-get install --quiet --assume-yes gridengine-qmon 
```
```
qconf -aq default  ## aggregate a new queue called "default"
qconf -mq default  ## modify the queue "default"
qconf -as localhost ## aggregate the localhost tho the list of submitters
```

Set the following values: `hostlist              localhost`

Take all default parameters BUT for the SGE master parameter, type `localhost` (it must be the hostname)

Test that jobs canbe sent to the job scheduler.


## 15. OPTIONAL

### Install some software tools for NGS analysis.
```
cd ${RSAT}
make -f makefiles/install_software.mk install_meme
```

### Ganglia: tool to monitor a cluster (or single machine) 
[Link.](https://www.digitalocean.com/community/tutorials/introduction-to-ganglia-on-ubuntu-14-04)
```
sudo apt-get install -y ganglia-monitor rrdtool gmetad ganglia-webfrontend
sudo cp /etc/ganglia-webfrontend/apache.conf /etc/apache2/sites-enabled/ganglia.conf
sudo apachectl restart
```

