---
title: "Install Snakemake workflows"
author: "Claire Rioualen"
date: "July 30, 2015"
output:
  html_document:
    css: ../styles.css
    fig_caption: yes
    toc: yes
    number_sections: true
  pdf_document:
    toc: yes
---

### **TODO**

* Include map of possible "bricks" of worflows (like general rulegraph) with each step's requirement/dependencies
* Include minimum json file config depending on bricks to be used.

* Beware of paths
    - ~/workspace
    - ~/bin
* Test all of this in IFB appliance

* **VBox issues**:
    - '/etc/init.d/vboxdrv setup' error when restarting
    - disparition vboxnet0
    
* Since snakemake 3.4: add pyYAML ?

**/!\\** attention pour les rsync, notamment si user + root... ssh, ssh agent ?

* revoir les install via apt-get car pb de version ! 
* mettre la procédure spécifique 
* lister les versions de chaque programme pour un wf qui fonctionne

* dependance mkvtree / rsat

* see differences between ubuntu and lmde (python libs notamment)
    
## Requirements

### VM creation

See one of these:

* `IFB_cloud_VM_creation.Rmd`
* `virtualbox_VM_creation.Rmd`

### VM customization

Make sure to provide the right amount of RAM and CPU, according to the amount of data, the tools you want to use, and what you actually have on your local installation.

Regarding the size of the virtual disk, see `virtualbox_VM_creation.Rmd`.

<!---
Here we used these settings:

* RAM: 40Go
* HD: 120Go
* CPU: 10
-->


# General requirements 

## Tools to install (in the event they're not)

### ssh
```
sudo apt-get install ssh
```
If you have followed one of the previous tutorials, you should have copied your ssh parameters onto your VM, and have an ssh public key like `~/.ssh/id_rsa.pub`. If not, you can generate keys using `ssh-keygen` (see manual [here](https://help.github.com/articles/generating-ssh-keys/)).

### rsync

rsync is an open source utility that provides fast incremental file transfer. 

* [rsync page](https://rsync.samba.org/)

```
sudo apt-get install rsync
```

### git

* Create an account on [GitHub](https://github.com).
* Install git on your machine.

```
sudo apt-get install git
```

* Add your ssh public key to your GitHub account settings (account > settings > SSH keys > add SSH key).

```
less ~/.ssh/id_rsa.pub
```

### zlib

Several tools require this dependency (e.g. sickle, bamtools...).
```
sudo apt-get install libz-dev
```

### qsub

**/?\\** TODO 
sudo apt-get install slurm-wlm-torque
! munge

## Create own bin

While some programs will be installed completely automatically... some will not. Here we create a directory that will be kept for manual installations. 
```
mkdir $HOME/bin
```
You might then have to edit your `$PATH` manually (see next section).

## Edit `$PATH`

In order to use manually installed programs and make them executable, you may have to update your `$PATH` environment variable. 
You can do so by editing the `~/.profile` file.

```
nano ~/.profile
```

Fetch this paragraph and add your new tool:

```
# set PATH so it includes user's private bin if it exists
if [ -d "$HOME/bin" ] ; then
    PATH="$HOME/bin/sickle:$PATH"
fi
```

Execute the file to validate the change. 

```
source ~/.profile
```

# Snakemake

/!\\ check R versions, python versions, etc...

/!\\ It seems there could an issue with the installation of python module while using sudo or doing is as root...
Googling "sudo pip" suggests it is a bad idea for safety reasons, and should not be needed...

## Python 3

Snakemake requires to have Python 3 installed. If you have installed Ubuntu14.04 on your machine by following on of the previous tutorials, you should have both Python 2.7 and Python 3.4 installed for starters. You can check this by issuing the following commands in a terminal: 
```
python --version # usually the default python version is 2.7+
python3 --version
```
If you don't have python 3 you should install it, as well as pip3.
```
sudo apt-get install python3
```
Install pip
```
sudo apt-get install python-pip
sudo apt-get install python3-pip
```
Not installed natively?
```
apt-get install python-dev
apt-get install python3.4-dev
```

## R

**/!\\* Issues with R lib 

Beware: libPath can be automatically set in rsat distribution, which can cause permission problems etc.
Check the first path in R:
```
> .libPaths()
```
Set your preferred path in shell:
``` 
$ export R_LIBS="/home/your_username/R_libs"
```

Issue when installing packages:

```
> setRepositories() 
```
Select CRAN, CRAN extras, bioc...

Beware of using "http" mirrors rather than "https" mirrors (why exactly?)

### Python 3 package rpy2

```
pip3 install "rpy2<2.3.10"
```

Issue with egg_info stuff, found info [here](http://stackoverflow.com/questions/17886647/cant-install-via-pip-because-of-egg-info-error)

```
doesnt work
```
Problems solved after rsat install (including loads of python packages)?

### spp (spp peak-calling)

depends on caTools
```
source("http://bioconductor.org/biocLite.R")
biocLite("spp")
> install.packages("spp")
```
```
R CMD INSTALL spp_1.10.tar.gz
```
...

### Biostrings (peak length)

```
install.packages("Biostrings", lib="/path/to/my/lib")
```

## Snakemake

* General documentation about snakemake: [click](https://bitbucket.org/johanneskoester/snakemake/wiki/Documentation)
* Snakemake FAQ/forum: [click](https://groups.google.com/forum/#!forum/snakemake)

Now you have installed Python 3 and pip3 (see previous section), you can install snakemake safely:
**/!\\** NB done as root#...
```
pip3 install snakemake
```
You can check snakemake works properly with this basic script:
```
"""Snakefile to test basic functions of snakemake.
"""
rule all:
	input: expand("bye.txt")

rule hello:
	"""Write HELLO in a text file named hello.txt.
	"""
	output: "hello.txt"
	message: "Generating {output} file."
	shell: "echo HELLO > {output}"

rule bye:
	"""Write BYE in a text file named bye.txt.
	"""
	input: "hello.txt"
	output: "bye.txt"
	message: "Generating {output} file."
	shell: "echo BYE > {output}"
```
* Save it to `~/workspace/hello.py`.
* Issue the command `cd ~/workspace ; snakemake -s hello.py`.
* 2 files shoudl be created: `hello.txt` and `bye.txt`.

**/!\\** Need version 3.4.1
```
pip3 install snakemake --upgrade
```

En option, pour générer les rapports:
```
pip3 install docutils
```

## Graphviz

Snakemake can generate useful graphviz outputs. 

```
sudo apt-get install graphviz
```


Now if we want to use snakemake to develop NGS workflows, we have to install a number of programs depending on what we want to achieve.

## Pandas

**/!\\ TODO** To be checked!

```
$ pip3 install pandas ## not ok
$ pip install pandas ## not ok
# pip install pandas ## not ok
# pip3 install pandas ## ok ?
```


# NGS analysis (optional) software to install

## SRA toolkit

This toolkit includes a number of programs, allowing the conversion of \*.sra files. `fastq-dump` translates \*.sra files to \*.fastq files.

* [SRA format](http://www.ncbi.nlm.nih.gov/Traces/sra/)
* [fastq-dump manual](http://www.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?view=toolkit_doc&f=fastq-dump)
* [Installation manual](http://www.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?view=toolkit_doc&f=std)

Download last version [here](http://www.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?view=software):
```
cd ~/bin
wget "http://ftp-trace.ncbi.nlm.nih.gov/sra/sdk/2.5.2/sratoolkit.2.5.2-ubuntu64.tar.gz"
tar -xvzf sratoolkit.2.5.2-ubuntu64.tar.gz
rm sratoolkit.2.5.2-ubuntu64.tar.g
```
 
Add to path (cf section 1.3): 
```
PATH="$HOME/bin/sratoolkit.2.5.2-ubuntu64/bin:$PATH"
```

Check version:
```
fastq-dump --version
  fastq-dump : 2.5.2
```


You should be able to install SRA toolkit simply by issuing this command, but likely it won't be the most recent release:
```
sudo apt-get install sra-toolkit
```

```
fastq-dump --version
  fastq-dump : 2.1.7
```
 

## FastQC

FastQC aims to provide a simple way to do some quality control checks on raw sequence data coming from high throughput sequencing pipelines. It provides a modular set of analyses which you can use to give a quick impression of whether your data has any problems of which you should be aware before doing any further analysis. 

* [FastQC main page](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/)

```
sudo apt-get install fastqc
```
V 0.10.1

Latest: 11.3


## Sickle

Sickle is a trimming tool which better the quality of NGS reads.

* [Sickle main page](https://github.com/najoshi/sickle)

* Pre-requisite: install `zlib` (see section 1.1.4).
* Clone the git repository into your bin (see section 1.2) and run `make`.

```
cd $HOME/bin
git clone https://github.com/najoshi/sickle.git
cd sickle
make
```

* Add sickle to your `$PATH` (see section 1.3).

```
PATH="$HOME/bin/sickle:$PATH"
```

## BWA

BWA is a software package for mapping low-divergent sequences against a large reference genome, such as the human genome.

* [Main page](http://bio-bwa.sourceforge.net/)
* [Manual](http://bio-bwa.sourceforge.net/bwa.shtml)

```
sudo apt-get install bwa
````
V: 0.7.5a-r405

Latest : 0.7.12

## Bowtie2

* [General documentation](http://bowtie-bio.sourceforge.net/bowtie2/manual.shtml)
* Instructions [here](http://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#obtaining-bowtie-2)
    * Download package [here](https://sourceforge.net/projects/bowtie-bio/files/bowtie2/)
    * Move package to your personnal bin/
    * Unzip
    * Add to $PATH (see section 1.3)
    * There you go!
    
```
cd ~/bin
wget http://sourceforge.net/projects/bowtie-bio/files/bowtie2/2.2.6/bowtie2-2.2.6-linux-x86_64.zip
unzip bowtie2-2.2.6-linux-x86_64.zip
```

## Samtools

SAM (Sequence Alignment/Map) format is a generic format for storing large nucleotide sequence alignments. 

* [SAMtools main page](http://samtools.sourceforge.net/)

```
sudo apt-get install samtools
```
V: 0.1.19
Latest: 1.2

## Bedtools

The bedtools utilities are a swiss-army knife of tools for a wide-range of genomics analysis tasks. For example, bedtools allows one to intersect, merge, count, complement, and shuffle genomic intervals from multiple files in widely-used genomic file formats such as BAM, BED, GFF/GTF, VCF. 

* [bedtools main page](http://bedtools.readthedocs.org/en/latest/)

```
sudo apt-get install bedtools
````
V: v2.17.0
Latest: 2.24.0

## MACS2

* [MACS2 web page](https://github.com/taoliu/MACS/)

```
sudo apt-get install python-numpy
sudo pip install MACS2
```


Marche pas:< 
```
$ git clone https://github.com/taoliu/MACS.git
# pip install MACS2
...
```

## HOMER

[Web page](http://homer.salk.edu/)

[Install instructions:](http://homer.salk.edu/homer/introduction/install.html)

wget "http://homer.salk.edu/homer/configureHomer.pl"
rsync -> VM
copy configureHomer.pl to /home/rg/bin/HOMER
cd /home/rg/bin/HOMER
perl configureHomer.pl -install homer

add to path : PATH="$HOME/bin/HOMER/bin:$PATH"


The basic Homer installation does not contain any sequence data.  To download sequences for use with homer, use the configureHomer.pl script.  To get a list of available packages:
perl /path-to-homer/configureHomer.pl -list
To install packages, simply use the -install option and the name(s) of the package(s).
perl /path-to-homer/configureHomer.pl -install mouse (to download the mouse promoter set)
perl /path-to-homer/configureHomer.pl -install mm8    (to download the mm8 version of the mouse genome)
perl /path-to-homer/configureHomer.pl -install hg19    (to download the hg19 version of the human genome)


Available genomes
Supported Organisms: Human (hg17, hg18, hg19), Mouse (mm8, mm9, mm10), Rat (rn4, rn5), Frog (xenTro2, xenTro3), Zebrafish (danRer7), Drosophila (dm3), C elegans (ce6, ce10), S. cerevisiae (sacCer2, sacCer3), pombe (ASM294v1), Arabidopsis (tair10), Rice (msu6), and also works with custom genomes in FASTA format and gene annotations in GTF format.


## SWEMBL

* [SWEMBL beginner's manual](http://www.ebi.ac.uk/~swilder/SWEMBL/beginners.html)

**TODO**


### R package spp

sudo su
echo "deb http://www.stats.bris.ac.uk/R/bin/linux/ubuntu precise/" >> /etc/apt/sources.list
apt-key adv --keyserver keyserver.ubuntu.com --recv-keys E084DAB9
apt-get update
apt-get upgrade

```
wget "https://cran.r-project.org/src/base/R-3/R-3.2.2.tar.gz"
tar -xf rm R-3.2.2.tar.gz
rm R-3.2.2.tar.gz
cd rm R-3.2.2
./configure
```
doesn't work on VM

not a problem of R version anyway

libboost libraries ?
apt-get install libboost-all-dev

```
> install.packages("caTools")
```


## RSAT suite

See *$RSAT/doc/howto/install_rsat_ubuntu14.04.bash*

# France Génomique ChIP-seq project - Scerevisiae-Pho4

## Cloning the repository

Cloning the repository from GitHub requires access rights. 
Contact: claire.rioualen@inserm.fr
```
cd ~/workspace
git clone git+ssh://git@git.renater.fr:2222/fg-chip-seq.git
```
Note: commit 3e1434c9942a2c7899b135f1a4be99d8f4aa68cf

## Data transfer

**/!\\ 15/08/31** Rsync ne fonctionne que dans un sens ? 
```
cd
mkdir data
mkdir genomes
```

Et depuis l'hôte:
```
rsync -ruptvl /data/Scerevisiae-Pho4 rg@192.54.201.107:~/data/
rsync -ruptvl /data/genomes/sc rg@192.54.201.107:~/genomes/
```


## Running the pipeline

```
cd ~/workspace/fg-chip-seq
snakemake -s scripts/snakefiles/workflows/Scerevisiae-Pho4.py -j 4
```

/!\\ path of files should be changed from /home/rioualen to /home/rg in both .py and .json files.

## Increase virtual disk space

You may need to expand your virtual disk space. For VBox users, see `virtualbox_vm_creation.Rmd`, section XX (VirtualBox hacks and tricks).

# VM export / submission 

See *$project_home/doc/install_protocols/export_appliance.Rmd*
