---
title: "France Genomique  Workpackage 2.6 - Gene Regulation"
author: "Claire Rioualen"
date: "March 17th, 2016"
---

<!--
Table of Contents
=================

  * [Purpose of this repository](#purpose-of-this-repository)
  * [Workflow examples](#workflow-examples)
    * [Transcription factors](#transcription-factors)
      * [Execution](#execution)
      * [Dependencies](#dependencies)
    * [Histone marks](#histone-marks)
      * [Execution](#execution-1)
      * [Dependencies](#dependencies-1)
    * [RNA-seq (TODO)](#rna-seq-todo)
  * [Data organisation](#data-organisation)
  * [Documentation](#documentation)
  * [Contact](#contact)
-->


This git repository holds shared code for the analysis of Next
Generation Sequencing data related to gene regulation: ChIP-seq,
RNA-seq, and related technologies. 

We have chosen to use the [Snakemake workflow management system](https://bitbucket.org/snakemake/snakemake/wiki/Home)[1] 
in order to build reproducible and flexible NGS analysis pielines.

# Snakemake workflows

Snakemake is a tool that enables the creation of analysis pipelines, based on the python language and the make concepts of rules and targets. 

A rule contains the commands to generate a given target. 
A workflow can be defined as a series of files generated by successive rules. 

Here, we present a repository of reusable NGS-specific rules, as well as a few workflow examples for standard ChIP-seq and RNA-seq analyses. 

# Workflow example

Our workflows are meant to work using two main files: a so-called workflow file (\*.py) and a configuration file (\*yml).
The first one contains the targets to be generated, and the second one contains the necessary and optional parameters 
to manage and customize the workflow. 

Here, we will show how to execute the [ChIP-seq workflow](https://github.com/rioualen/gene-regulation/blob/master/scripts/snakefiles/workflows/factor_workflow.py) `factor_workflow.py`.

In the [example directory](https://github.com/rioualen/gene-regulation/blob/master/examples) you will find several case studies. We will use the `GSE20870` yeast study. It comes with three tab-delimited files: 

* `merge.tab` contains the information for merging lanes;
* `samples.tab` contains a description of each samples;
* `design.tab` contains the samples to be compared. 

The `yml` configuration file contains the paths to the data to be analysed, as well as the data to be generated. 
Here, we analyse raw samples (\*.sra) located in the `/data` directory: 

```
/data/
    GSE20870/
        GSM521934/
            SRR051929.sra
        GSM521935/
            SRR051930.sra
```
Raw data is downloadable from the GEO platform. The automatic downloading is soon to be integrated in the workflow. 

The reference genome will be automatically downloaded in the `/data/genome` directory:

```
/data/
    genomes/
        sacCer2/
            sacCer2.fa
```

You should adapt the configuration file required fields to your own file organization. 

# Workflow execution

The workflow includes the configuration file using this code:
```
configfile: "examples/GSE20870/GSE20870.yml"
```
You can then chose what program(s) to use for mapping and peak-calling by commenting/uncommenting the following lines:

```
ALIGNER=[
#    "bowtie",
    "bowtie2",
#    "bwa"
]
```
```
PEAKCALLER=[
    "homer-fdr" + config["homer"]["fdr"] + "_peaks", 
    "macs2-qval" + config["macs2"]["qval"], 
    "swembl-R" + config["swembl"]["R"],
#    "macs14-pval" + config["macs14"]["pval"],
#    "spp-fdr" + config["spp"]["fdr"],
#    "bPeaks_allGenome"
]
```

You can now run the workflow:
```
snakemake -s scripts/snakefiles/workflows/factor_workflow.py
```

The workflow generates a flowchart of the analysis:

![alt text][factor]


# Purpose of this repository

This git repository holds shared code for the analysis of Next
Generation Sequencing data related to gene regulation: ChIP-seq,
RNA-seq, and related technologies.

The goals are multiple.

1. Avoid duplication of efforts by sharing the developments of
different bioinformaticians involved in ChIP-seq and RNA-seq projects.

2. Ensure portability and re-usability of the code.

3. Enable validation of the code by independent users.


# Pre-requisites

In order to run the workflow as shown above, you should have the following programs installed:

* R 3+
* Python 3.3+
* Snakemake 3.3+

* [SRA Toolkit](http://www.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?view=software)
* [Sickle](https://github.com/najoshi/sickle)
* [Bowtie 2](http://bowtie-bio.sourceforge.net/)
* [SAMtools 1.3+](http://samtools.sourceforge.net/)
* [FastQC](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/)
* [bedtools](http://bedtools.readthedocs.org/)
* [HOMER](http://homer.salk.edu/homer/index.html)
* [MACS2](https://github.com/taoliu/MACS/)
* [SWEMBL](http://www.ebi.ac.uk/~swilder/SWEMBL/)
* [RSAT](http://rsat.eu/)



See `doc/install_protocols/install_snakemake_workflows.Rmd` for a detailed manual on how to install programs and dependencies.
A makefile will soon be available to perform the installation of dependencies.

# Documentation

More documentation can be found in the `doc` directory. (under construction)

It includes: 

* Instructions for the installation of dependencies (to be refhreshed)
* RSAT install guide (to be refreshed)
* Instructions for building a virtual machine on the IFB cloud or under VirtualBox
* Some wiki material



# Contact

- Claire Rioualen <claire.rioualen@inserm.fr>
- Jacques van Helden <Jacques.van-helden@univ-amu.fr>

# References 

1. KÃ¶ster, Johannes and Rahmann, Sven. "Snakemake - A scalable bioinformatics workflow engine". Bioinformatics 2012.

[factor]: https://github.com/rioualen/gene-regulation/blob/master/examples/chip-seq_factor.png