# Check if each required parameter has been defined in the JSON
# configuration file, otherwise use default values

if not "bwa" in config.keys():
    config["bwa"] = {}

if not "index" in config["bwa"].keys():
#    config["bwa"]["index"] = ""
    sys.exit("The parameter bwa index must be defined in the JSON file")

if not "threads" in config["bwa"].keys():
    config["bwa"]["threads"] = "1"

rule bwa_se:
	"""Read mapping using BWA. 

	Requires the indexing to have previously been done (using the
	rule bwa_index).  This is checked by testing the presence of
	file {genome}.done

	Parameters:
		config["bwa"]["index"]
		config["bwa"]["threads"]
		config["qsub"]

	Usage:
		input:expand("{dataset}/{dataset}_bwa.sam", dataset=DATASET)

        """
	input: 	reads="{reads}.fastq", \
		index=config["bwa"]["index"] + ".bwt"
	output: "{reads}_bwa.sam"
	log: "{reads}_bwa.log"
	benchmark: "{reads}_bwa_benchmark.json"
	params:
		index = config["bwa"]["index"], \
		sai= "{reads}_bwa.sai", \
		qsub = config["qsub"] + " -e {reads}_bwa_qsub.err -o {reads}_bwa_qsub.out", \
		threads = config["bwa"]["threads"]
	shell:
		"bwa aln -t {params.threads} {params.index} {input.reads} > {params.sai} 2> {log}; \
		bwa samse {params.index} {params.sai} {input.reads} > {output} 2> {log}"
