"""Read mapping for single end data using BWA. 
Requires the indexing to have previously been done (using the
rule bwa_index).

Required parameters:
    config["genome"]["version"]
    config["qsub"]

Optional parameters:
    config["bwa"]["dir"]
    config["bwa"]["threads"]

Usage example:


Contributors: 
    Claire Rioualen
"""

# Set parameters
if not "qsub" in config.keys():
    sys.exit("The parameter qsub must be defined in the config file")

if not "version" in config["genome"].keys():
    sys.exit("Genome version must be defined in the config file")

if not "genome" in config["dir"].keys():
    sys.exit("Genome dir must be defined in the dir section of the config file")

if not "bwa" in config.keys():
    config["bwa"] = {}

if not "threads" in config["bwa"].keys():
    config["bwa"]["threads"] = "1"

# Define paths
if not "samples" in config["dir"].keys():
    config["dir"]["samples"] = config["dir"]["results"]

genome_version = config["genome"]["version"]
genome_dir = config["dir"]["genome"] + genome_version + "/"
genome_file = genome_dir + genome_version + ".fa"
bwa_dir = genome_dir + "bwa/"
bwa_index = bwa_dir + genome_version + ".fa"


rule bwa_se:
    input:
        reads = "{reads}.fastq", \
        index = bwa_index
    output:
        sam = "{reads}_bwa.sam"
    params:
        threads = config["bwa"]["threads"], \
        sai= "{reads}_bwa.sai", \
        qsub = config["qsub"] \
          + " -e " + "{reads}_bwa_qsub.err" \
          + " -o " + "{reads}_bwa_qsub.out"
    log: "{reads}_bwa.log"
    benchmark: "{reads}_bwa_benchmark.json"
    message: "Mapping reads on reference genome {input.index} using BWA..."
    shell:"""
bwa aln -t {params.threads} {input.index} {input.reads} > {params.sai} 2> {log}
bwa samse {input.index} {params.sai} {input.reads} > {output} 2> {log}
"""







