"""Snakemake rules to download a reference genome from Ensembl.

Sequences (fasta format) + genomic annotations (gtf format).
"""
## Attention, Jacques suppressed the gff3 download, because it was not
## working with E.coli, and apparently we don't use these files, we
## use the gtf instead.

rule bowtie2_download_genome:
    """Download and index the reference genome for the Bowtie2
    read-mapping software from ensembl.


    """
    # output: sequence = config["dir"]["genome"] + "/" + config["genome"]["sequence"] + ".fa", \
    #         features_gtf = config["dir"]["genome"] + "/" + config["genome"]["features_gtf"] + ".gtf"
    # log: config["dir"]["genome"] + "/" + "wget_log.txt"
    # params: sequence_url = config["genome"]["base_url_fasta"] + config["genome"]["sequence"] + ".fa.gz", \
    #         features_gtf_url = config["genome"]["base_url_gtf"] + config["genome"]["features_gtf"] + ".gtf.gz"
    output: sequence = config["dir"]["genome"] + "/" + config["genome"]["sequence"] + ".fa", \
            features_gtf = config["dir"]["genome"] + "/" + config["genome"]["features_gtf"] + ".gtf"
    log: config["dir"]["genome"] + "/" + "wget_log.txt"
    params: sequence_url = config["genome"]["base_url_fasta"] + "/" + config["genome"]["sequence"] + ".fa.gz", \
            features_gtf_url = config["genome"]["base_url_gtf"] + "/" + config["genome"]["features_gtf"] + ".gtf.gz", \
            qsub = config["qsub"]
    shell: "wget --timestamping --no-directories " \
        + " --directory-prefix " + config["dir"]["genome"] + "/" \
        + " {params.sequence_url}" \
        + " {params.features_gtf_url}" \
        + " && gunzip -f " \
        +  config["dir"]["genome"] + "/" + config["genome"]["sequence"] + ".fa.gz " \
        +  config["dir"]["genome"] + "/" + config["genome"]["features_gtf"] + ".gtf.gz " \
        + " &> {log}"
