"""Compute genome coverage from aligned reads.

Requirements: 

- Kent userApps (http://hgdownload.cse.ucsc.edu/admin/exe/userApps.src.tgz)
- bedtools
"""

rule install_bedgraphtobigwig:
    params: url="http://hgdownload.cse.ucsc.edu/admin/exe/userApps.src.tgz"
    shell: "wget {params.url}"


# ## Set defaults params in case they are not defined in config file. 
# if not "genome_coverage" in config.keys():
#     config["genome_coverage"] =  {'intervals': 200}

rule genome_coverage_dz:
    """Compute coverage (reads per position) for each position of a
    genome, from a bam-formatted file with aligned reads.

    BEWARE: this rule is usefull for small genomes (Bacteria, Fungi)
    but would produce a very big file for Metazoan or Plant genomes.

    """
    input: reads="{reads}.bam"
    output: "{reads}_genomecov_dz.txt"
    log:  "{reads}_genomecov_dz.log"
    benchmark:  "{reads}_genomecov_dz_benchmark.json"
    params: qsub = config["qsub"] + " -e {reads}_genomecov_dz_qsub.err -o {reads}_genomecov_dz_qsub.out"
    shell: "bedtools genomecov -strand -dz -ibam {input.reads} " \
        + "  > {output} 2> {log}"

rule genome_coverage_bedgraph:
    """Compute genome coverage from a bam-formatted file with aligned
    reads.  The coverage file is in bedgraph format (with extension
    .bedgraph), which can be loaded in the genome viewer IGV. 

    Note however that IGV issues a warning when bedgraph files are
    given in input, and recommends to use the tdf format instead. We
    implemented hereafter rules to convert bedgraph to tdf.

    """
    input: reads = "{reads}.bam"
    output: bedg = "{reads}.bedgraph"
    log:  "{reads}_genomecov_bedgraph.log"
    benchmark:  "{reads}_genomecov_bedgraph_benchmark.json"
    params: qsub = config["qsub"] + " -e {reads}_genomecov_bedgraph_qsub.err -o {reads}_genomecov_bedgraph_qsub.out"
    shell: "bedtools genomecov -bg -ibam {input.reads} > {output.bedg} 2> {log}" 

# rule genome_coverage_bedgraph2:
#     """Compute genome coverage from a bam-formatted file with aligned
#     reads.  The coverage file is in bedgraph format, which can be
#     loaded in the genome viewer IGV.

#     """
#     input: reads = "{reads}.bam"
#     output: bedg = "{reads}_genomecov.bedgraph"
#     log:  "{reads}_genomecov_bedgraph.log"
#     benchmark:  "{reads}_genomecov_bedgraph_benchmark.json"
#     params: qsub = config["qsub"] + " -e {reads}_genomecov_bedgraph_qsub.err -o {reads}_genomecov_bedgraph_qsub.out"
#     shell: "bedtools genomecov -bg -ibam {input.reads} > {output.bedg} 2> {log}" 

rule genome_coverage_bedgraph_strands:
    """Compute two strand-specific genome coverage files (with suffixes
    _strand- and _strand+) from a bam-formatted file with aligned
    reads.  The coverage files are in bedgraph format, which can be
    loaded in the genome viewer IGV.

    """
    input: reads = "{reads}.bam"
    output: bedg_plus = "{reads}_genomecov_strand+.bedgraph", \
            bedg_minus = "{reads}_genomecov_strand-.bedgraph"
    log:  "{reads}_genomecov_bedgraph_str.log"
    benchmark:  "{reads}_genomecov_bedgraph_str_benchmark.json"
    params: qsub = config["qsub"] + " -e {reads}_genomecov_bedgraph_str_qsub.err -o {reads}_genomecov_bedgraph_str_qsub.out"
    shell: "bedtools genomecov -bg -strand + -ibam {input.reads} > {output.bedg_plus} 2> {log}" \
        + "; bedtools genomecov -bg -strand - -ibam {input.reads} > {output.bedg_minus} 2>> {log}"

rule bedgraph_to_tdf:
    """Convert bedgraph to TDF format, which is recommended to load
    coverage data in IGV.
    
    The conversion relies on igvtools:
    https://www.broadinstitute.org/software/igv/igvtools

    """
    input: "{coverage}.bedgraph"
    output: "{coverage}.tdf"
    log: "{coverage}_bedgraph_to_tdf.log"
    benchmark: "{coverage}_bedgraph_to_tdf_benchmark.json"
    params: genome=config["genome"]["fasta"], \
            qsub = config["qsub"] + " -e {coverage}_bedgraph_to_tdf_qsub.err -o {coverage}_bedgraph_to_tdf_qsub.out"
    shell: "igvtools totdf {input} {output} {params.genome} 2> {log}"


## !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
## TEMPORARILY COMMENTED BECAUSE IT REQUIRES
## config["genome"]["chromsize"] for which we have no solution yet.
#rule bam2wig:    
#    """Compute genome coverage from a bam-formatted file with aligned
#    reads and produce a wig or bigWig file, the recommended format to
#    upload coverage-type data as UCSC tracks.

#    """
#    input: "{reads}.bam"
#    output: "{reads}.wig"
#    log:  "{reads}_bam2wig.log"
#    benchmark:  "{reads}_bam2wig_benchmark.json"
#    params: output_prefix = "{reads}", \
#            chromsize = config["genome"]["chromsize"], \
#            qsub = config["qsub"] + " -q long -e {reads}_bam2wig_qsub.err -o {reads}_bam2wig_qsub.out"
#            qsub = config["qsub"] + " -e {reads}_bam2wig_qsub.err -o {reads}_bam2wig_qsub.out"
#    shell: "bam2wig.py -i {input} -s {params.chromsize} -o {params.output_prefix} 2> {log}" 
## !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!


## rule bedgraph_to_bigwig:
# TO BE WRITTEN    

## rule bam_to_tdf:
# TO BE WRITTEN, TAKING INTO ACCOUNT THE STRAND
# igvtools help count  to find the proper way

rule bedgraphtobigwig:
    """Convert bedGraph to bigWig format."""
    input: chromizes="", bedgraph="{coverage}.bedGraph"
    output: "{coverage}.bw"
    log:  "{coverage}_bedGraphToBigWig.log"
    benchmark:  "{coverage}_bedGraphToBigWig.json"
    params: qsub = config["qsub"] + " -e {coverage}_bedGraphToBigWig_qsub.err -o {coverage}_bedGraphToBigWig_qsub.out"
    shell: "bedGraphToBigWig {input.bedgrap} {input.chromsizes} {output} 2> {log}"

# rule genome_coverage_bigwig:
#     """Compute genome coverage from a bam-formatted file with aligned reads.

#     Parameters:
#     config["genome"]["chromsizes"] File describing chromosome sizes
    
#     """
#     input: reads="{reads}.bam"
#     output: bw="{reads}_genomecov.bw"
#     log:  "{reads}_genomecov.log"
#     benchmark:  "{reads}_genomecov_benchmark.json"
#     params: chrom_sizes = config["genome"]["chromsizes"], \
#             intervals = config["genome_coverage"]["intervals"], \
#             qsub = config["qsub"] + " -e {reads}_genomecov_qsub.err -o {reads}_genomecov_qsub.out"
#     shell: "bedtools genomecov -bg -scale {params.intervals} -ibam {input.reads} " \
#         + " | bedGraphToBigWig  > {output.bw} 2> {log}"


