# Set parameters
if not "qsub" in config.keys():
    sys.exit("The parameter qsub must be defined in the config file")

# Rule
rule genome_coverage_dz:
    """Compute coverage (reads per position) for each position of a
    genome, from a bam-formatted file with aligned reads.

    BEWARE: this rule is useful for small genomes (Bacteria, Fungi)
    but would produce a very big file for Metazoan or Plant genomes.

    Required parameters:
        config["qsub"]

    Copyright Jacques van Helden
    """
    input: reads="{reads}.bam"
    output: "{reads}_genomecov_dz.txt"
    params: qsub = config["qsub"] + " -e {reads}_genomecov_dz_qsub.err -o {reads}_genomecov_dz_qsub.out"
    log:  "{reads}_genomecov_dz.log"
    benchmark:  "{reads}_genomecov_dz_benchmark.txt"
    shell: "bedtools genomecov -strand -dz -ibam {input.reads} > {output} 2> {log}"

