rule macs2:
	"""Peak-calling with MACS2.
	Input: sam, bam, bed... automatically detected by macs' callpeak algorithm. Bed format was used here. 
	Output: bed

	To be refined...
	TODO:
		Add qsub, log, bench, message keywords.
		Generalize rule... And clean up this mess of a file once it's done.

	Required parameters:
		config['qsub']
		config['macs2']['genome']
		config['dir']['results']

	Usage: 
		MACS2 = expand(RESULTSDIR + "{chip}_vs_{inp}/{chip}_vs_{inp}_{trimming}_{aligner}_{caller}_peaks.xls", chip=CHIP, inp=INPUT, trimming=TRIMMING, aligner=ALIGNER, caller=PEAK_CALLER)
	"""
        input: treatment="{result_dir}/{treatment}/{treatment}" + config["suffixes"]["mapped"] + ".bed", \
            control="{result_dir}/{control}/{control}" + config["suffixes"]["mapped"] + ".bed"
	output: "{result_dir}/{treatment}_vs_{control}/{treatment}_vs_{control}" + config["suffixes"]["mapped"] + "_macs2_summits.bed"
        params: outdir="{result_dir}/{treatment}_vs_{control}", \
            name="{treatment}_vs_{control}" + config["suffixes"]["mapped"] + "_macs2._summits",\
            genome_size = config["genome"]["size"]
	shell: "macs2 callpeak -t {input.treatment} -c {input.control} --gsize {params.genome_size} --outdir {params.outdir} --name {params.name}"

